File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/G._Максимальный_путь_в_дереве/solution.cpp
#ifndef REMOTE_JUDGE
struct Node {
  int value;
  const Node* left = nullptr;
  const Node* right = nullptr;
  Node(int value, Node* left, Node* right) : value(value), left(left), right(right) {}
};
#endif

#ifdef REMOTE_JUDGE
#include "solution_tree.h"
#endif
#include <cmath>
#include <iostream>
#include <cassert>
#include <cstdint>
using namespace std;


int dfs(const Node* root, int& maxSum) {
    if (!root) return INT32_MIN;
    int leftSum = std::max(dfs(root->left, maxSum), 0);
    int rightSum = std::max(dfs(root->right, maxSum), 0);
    maxSum = std::max(
        maxSum, root->value + leftSum + rightSum
    );
    return root->value + std::max(leftSum, rightSum);
}


int Solution(const Node* root) {
    /*
    Time Complexity O(N)
    Memory Complexity O(H)
    */
    int maxSum = INT32_MIN;
    dfs(root, maxSum);
    return maxSum;
}

#ifndef REMOTE_JUDGE
void test() {
    Node node1({5, nullptr, nullptr});
    Node node2({1, nullptr, nullptr});
    Node node3({-3, &node2, &node1});
    Node node4({2, nullptr, nullptr});
    Node node5({2, &node4, &node3});
    assert(Solution(&node5) == 6);
}

int main() {
  test();
}
#endif

File: Yandex/Practicum/Algorightms_and_Data_Structures/5_sprint/G._Максимальный_путь_в_дереве/solution.py
import os

LOCAL = os.environ.get('REMOTE_JUDGE', 'false') != 'true'

if LOCAL:
    class Node:
        def __init__(self, value, left=None, right=None):
            self.value = value
            self.right = right
            self.left = left


def solution(root: 'Node') -> int:
    """
    Time Complexity O(N)
    Memory Complexity O(H)
    """
    max_sum: int = float('-inf')
    def dfs(root: 'Node') -> int:
        if not root: return float('-inf')
        left_sum: int = max(dfs(root.left), 0)
        right_sum: int = max(dfs(root.right), 0)
        nonlocal max_sum
        max_sum = max(
            max_sum, root.value + left_sum + right_sum
        )
        return root.value + max(left_sum, right_sum)
    dfs(root)
    return max_sum


def test():
    node1 = Node(5, None, None)
    node2 = Node(1, None, None)
    node3 = Node(-3, node2, node1)
    node4 = Node(2, None, None)
    node5 = Node(2, node4, node3)
    assert solution(node5) == 6


if __name__ == '__main__':
    test()